### ISSUE TITLE
Add support for analysing SMTP TLS reports

### ISSUE THREAD

**Issue by ghost** [domainaware/parsedmarc#71]
Hi

Have you planed that your tool can analyze the report from mta-sts (TLSRPTv1) too?

Or do you know another software for this?

Thank you for help

**Comment by thafreeman** [comment]
[URIports](https://www.uriports.com) supports MTA-STS and DANE TLS-RPT reports.

**Comment by seanthegeek** [comment]
I just glanced over [the RFC](https://tools.ietf.org/html/rfc8460#section-4). This looks like it would be easy to add. Not sure when I'll get to it though.

**Comment by dragoangel** [comment]
@freddieleeman all point in self-hosted solutions is that they self-hosted. You not share your personal info with 3rd parties when you not want too.

**Comment by dragoangel** [comment]
Hi @seanthegeek I can send you sample data for mta-sts and tlsa reports if it will help you. Do you need them?

**Comment by seanthegeek** [comment]
@dragoangel That would be great!

**Comment by dragoangel** [comment]
@seanthegeek I contact you in PM on twitter

**Comment by NoSubstitute** [comment]
> @dragoangel That would be great!

Do you need more reports? I'm getting some incoming, and no clue how to read them. :-)

**Comment by seanthegeek** [comment]
I haven't even had the chance to read over what @dragoangel sent me. 😋 I'll keep that in mind though. Thanks.

**Comment by dragoangel** [comment]
> @NoSubstitute no clue how to read them
TLS Reports are JSON formated mostly to one line. To get there more visibility you need format them  to pretty json (using online tools, or text app, e.g: Notepadd++ JSTool plugin, etc).

**Comment by NoSubstitute** [comment]
> > @NoSubstitute no clue how to read them
> 
> TLS Reports are JSON formated mostly to one line. To get there more visibility you need format them to pretty json (using online tools, or text app, e.g: Notepad++ JSTool plugin, etc).

Thanks. The NP++ plugin made it easier to read. Still, a nicely aggregated statistical view would be nice.

**Comment by rhclayto** [comment]
It would be a splendid feature. Is anyone aware of a currently existing self-hosted TLS-RPT analyzer/dashboard?

**Comment by dragoangel** [comment]
If you speak about open source - no, I know parsedmarc project that parse dmarcs, and it would be cool to add there tls reports, but as far I know it not implemented.

**Comment by tbsmark86** [comment]
I've done a quick solution for this writing **only** to json output; so no analyzing or anything just re-using the  automatic imap-mailbox handling. (It's enough for my use case)

Maybe some on can use this as a starting point for a real implementation:
https://github.com/tbsmark86/parsedmarc/commit/990f8df60baa017fcd5aad9871c00f70f3dcafc7

**Comment by mapietru** [comment]
+1 for that feature.

**Comment by bbccdd** [comment]
+1 !

**Comment by cbrandlehner** [comment]
+1 for MTA-STS support

**Comment by ArenTahmasian** [comment]
+1 for MTA-STS

**Comment by matthiasmaes** [comment]
+1 for MTA-STS

**Comment by Zoey2936** [comment]
Is there something new on this? I've discovered this project today and for my use case is this is the only feature, which I miss.

**Comment by mkilijanek** [comment]
+1 !

**Comment by ermurenz** [comment]
+1 for MTA-STS support

**Comment by Pascal76** [comment]
+1 :)

**Comment by Kuzuto** [comment]
To those who want TLS Reporting.. Have you set it up, and receiving reports ?
Can I get some replies from who you receive TLS Reports from ?

I can start out:

- Google

**Comment by freddieleeman** [comment]
- Google Inc.
- Microsoft Corporation
- SocketLabs
- Comcast
- Mail.ru
- Mimecast

And an additional 20 that lack significance.

**Comment by cbrandlehner** [comment]
I agree that google.com and microsoft.com are majority of emails here.

However, adding the following:
- yahoo.de
- yahoo.co.uk
- yahoo.com
- verizon.net

**Comment by freddieleeman** [comment]
Are you sure? Haven't seen a single report, and we process thousands daily.

**Comment by mkilijanek** [comment]
SMTP-TLS reports I receive are from Google.

**Comment by Kuzuto** [comment]
> * Google Inc.
> 
>     * Microsoft Corporation
> 
>     * SocketLabs
> 
>     * Comcast
> 
>     * Mail.ru
> 
>     * Mimecast
> 
> 
> And an additional 20 that lack significance.

It'a amazing so many sends TLS Reports now. Only a few years ago, only 4 was sending reports, where Microsoft was the last coming to the group. : https://www.mailhardener.com/blog/microsoft-has-begun-sending-smtp-tls-reports
Never got any from Mimecast. Would like to see that report.
Is the TLS reports you receive from all those different senders RFC Compliant, or is any still lacking behind ?

**Comment by jeffrysleddens** [comment]
In order of magnitude:

- google
- microsoft
- socketlabs
- comcast
- mail.ru
- alogica.it
- lethe-nieuwburg.nl
- net-centre.net
- keysys.com
- peetersgroup.com


I have almost 2000 SMTP TLS reports saved up.

**Comment by zell-mbc** [comment]
Small business email server:
- google
- microsoft

**Comment by wblondel** [comment]
I receive TLS reports from:

- google.com
- mail.ru
- microsoft.com

**Comment by Ressy66** [comment]
TLS-RPT will never take off because nowhere can I find information on how to generate and SEND reports, nobody else I knows, knows either, so its only half baked idea with results if you get mail from google and microsoft, nothing from the 50 million other mail servers out there

**Comment by dragoangel** [comment]
> TLS-RPT will never take off because nowhere can I find information on how to generate and SEND reports, nobody else I knows, knows either, so its only half baked idea with results if you get mail from google and microsoft, nothing from the 50 million other mail servers out there

https://datatracker.ietf.org/doc/html/rfc8460 it has own rfc thing, how you can say after that it's half baked? If you like details - read rfc :) . Fact that there no much support (only ~10 big providers), yes, but not 2. It's because there no open source software that support parsing it or sending it.

**Comment by Kuzuto** [comment]
> TLS-RPT will never take off because nowhere can I find information on how to generate and SEND reports, nobody else I knows, knows either, so its only half baked idea with results if you get mail from google and microsoft, nothing from the 50 million other mail servers out there

I kind of agree with you. TLS-RPT is easy to implement for the domain owner, but to get the full use of it, your mail service provider need to support MTA-STS or DANE as well. 
And this is something many mail service providers don't even support yet! Hence the reason we don't get reports..
Not everybody is only sending and receiving messages between Google and Microsoft..
Both the sending and receiving part, needs to support MTA-STS to prevent message from getting delivered, even when TLS fails.
And when TLS fails .. well.. you get a TLS-RPT within the next 24 hours telling you, "Google was unable to deliver messages to you, because TLS was failing". No company will accept to get this information up to 23:59 hours later. So what is the point ?
Many mail service providers is able to define inbound TLS rules and versions, to ensure TLS delivery. It's even possible to do exceptions.. That is not possible with MTA-STS.

Just like BIMI.. sounds good on paper, but all the big players have removed their BIMI records. It is expensive to buy certificate, and when only a few mail clients support it, and required by the MTA servers .. it dies out.

Microsoft support MTA-STS (last year) https://techcommunity.microsoft.com/t5/exchange-team-blog/introducing-mta-sts-for-exchange-online/ba-p/3106386
#NOTE: Messages will be delivered when only one party supports MTA-STS. For example, when an MTA-STS-protected domain receives a message from a sender domain that doesn’t support MTA-STS, the message is delivered. The message is also delivered when the recipient domain doesn’t support MTA-STS, but the sender domain does. The only scenario where messages aren’t delivered is when both sides are using MTA-STS and MTA-STS validation fails.

Again from : https://www.mailhardener.com/kb/smtp-tls-reporting
#So, the SMTP session always starts in plain text, and only switches to encrypted communication after the STARTTLS command is issued. Because TLS can fail in more ways than plain text connections, most MTAs will fall back to plain text if for some reason the TLS connection does not work.
With the introduction of [MTA-STS], it is now possible to enforce the use of TLS when receiving email. The plain-text fallback will no longer be allowed, so if TLS fails the email will not be delivered and returned to its sender.

I only see a change of TLS-RPT use, when more MTA servers support MTA-STS or DANE. And they need to support both.. because both is possible to enforce TLS connections. Else, what is the report good for?
Please enlighten me, if TLS-RPT is a company critical thing.

**Comment by seanthegeek** [comment]
I've finally started working on this, as you can see in PR #453. Lots of work still needs to be done, so I would appreciate any help!

**Comment by ermurenz** [comment]
Most of SMTP-TLS reports we receive are from Google.com and microsoft.com.Some other more from mail.ru.
